packages/sdk-js/src/plugins/growthbook-tracking.ts

```diff
@@ -0,0 +1,289 @@
+import type {
+  GrowthBookClient,
+  UserScopedGrowthBook,
+} from "../GrowthBookClient";
+
+const SDK_VERSION = loadSDKVersion();
+
+type GlobalTrackedEvent = {
+  eventName: string;
+  properties: Record<string, unknown>;
+};
+declare global {
+  interface Window {
+    gbEvents?:
+      | (GlobalTrackedEvent | string)[]
+      | {
+          push: (event: GlobalTrackedEvent | string) => void;
+        };
+  }
+}
+
+type EventPayload = {
+  event_name: string;
+  properties_json: Record<string, unknown>;
+  sdk_language: string;
+  sdk_version: string;
+  url: string;
+  context_json: Record<string, unknown>;
+  user_id: string | null;
+  device_id: string | null;
+  page_id: string | null;
+  session_id: string | null;
+  page_title?: string;
+  utm_source?: string;
+  utm_medium?: string;
+  utm_campaign?: string;
+  utm_term?: string;
+  utm_content?: string;
+};
+
+function parseString(value: unknown): null | string {
+  return typeof value === "string" ? value : null;
+}
+
+function parseAttributes(
+  attributes: Attributes
+): {
+  nested: Attributes;
+  topLevel: {
+    user_id: string | null;
+    device_id: string | null;
+    page_id: string | null;
+    session_id: string | null;
+    page_title?: string;
+    utm_source?: string;
+    utm_medium?: string;
+    utm_campaign?: string;
+    utm_term?: string;
+    utm_content?: string;
+  };
+} {
+  const {
+    user_id,
+    device_id,
+    anonymous_id,
+    id,
+    page_id,
+    session_id,
+    utmCampaign,
+    utmContent,
+    utmMedium,
+    utmSource,
+    utmTerm,
+    pageTitle,
+    ...nested
+  } = attributes;
+
+  return {
+    nested,
+    topLevel: {
+      user_id: parseString(user_id),
+      device_id: parseString(device_id || anonymous_id || id),
+      page_id: parseString(page_id),
+      session_id: parseString(session_id),
+      utm_campaign: parseString(utmCampaign) || undefined,
+      utm_content: parseString(utmContent) || undefined,
+      utm_medium: parseString(utmMedium) || undefined,
+      utm_source: parseString(utmSource) || undefined,
+      utm_term: parseString(utmTerm) || undefined,
+      page_title: parseString(pageTitle) || undefined,
+    },
+  };
+}
+
+type EventData = {
+  eventName: string;
+  properties: EventProperties;
+  attributes: Attributes;
+  url: string;
+};
+
+async function track({
+  clientKey,
+  ingestorHost,
+  events,
+}: {
+  events: EventData[];
+  clientKey: string;
+  ingestorHost?: string;
+}) {
+  if (!events.length) return;
+  const data: EventPayload[] = events.map(
+    ({ eventName, properties, attributes, url }) => {
+      const { nested, topLevel } = parseAttributes(attributes || {});
+
+      return {
+        event_name: eventName,
+        properties_json: properties || {},
+        ...topLevel,
+        sdk_language: "js",
+        sdk_version: SDK_VERSION,
+        url: url,
+        context_json: nested,
+      };
+    }
+  );
+
+  const endpoint = `${
+    ingestorHost || "https://us1.gb-ingest.com"
```

Comment:
Probably out of scope, but do we need to think about DNS level killswitches here for DDOS?

Reply:
Since this will be tied to billing on day one, I'm less worried about it.
This ingestion domain will probably end up on adblock lists eventually.  So we will likely offer a custom CNAME option for bigger enterprises.